legion/engine/physics/mesh_splitter_utils/mesh_splitter.hpp

```diff
@@ -0,0 +1,749 @@
+       */
+        void InitializePolygons(ecs::entity_handle entity)
+        {
+            owner = entity;
+
+            auto [meshFilter, meshRenderer] = entity.get_component_handles<rendering::mesh_renderable>();
+
+            ownerMaterialH = meshRenderer.read().material;
+
+            auto [posH, rotH, scaleH] = entity.get_component_handles<transform>();
+
+            if (meshFilter && posH && rotH && scaleH)
+            {
+                log::debug("Mesh and Transform found");
+                std::queue<meshHalfEdgePtr> meshHalfEdges;
+
+                //auto renderable = renderable.read();
+                mesh& mesh = meshFilter.read().get().second;
+
+                const math::mat4 transform = math::compose(scaleH.read(), rotH.read(), posH.read());
+
+                HalfEdgeFinder edgeFinder;
+                edgeFinder.FindHalfEdge(mesh, transform, meshHalfEdges);
+
+                BFSPolygonize(meshHalfEdges, transform);
+
+                log::debug("Mesh vertices {}, Mesh indices {}", mesh.vertices.size(), mesh.indices.size());
+
+            }
+            else
+            {
+                log::warn("The given entity does not have a meshHandle!");
+            }
+
+        }
+
+        /** @brief Given a queue of edges and a transform,
+        * populates the std::vector 'meshPolygons' using BFS
+        */
+        void BFSPolygonize(std::queue<meshHalfEdgePtr>& halfEdgeQueue, const math::mat4& transform)
+        {
+            //while edge queue is not empty
+            while (!halfEdgeQueue.empty())
+            {
+                meshHalfEdgePtr startEdge = halfEdgeQueue.front();
+                halfEdgeQueue.pop();
+
+                if (!startEdge->isVisited)
+                {
+                    SplittablePolygonPtr polygon = nullptr;
+
+                    if (BFSIdentifyPolygon(startEdge, polygon, halfEdgeQueue, transform))
+                    {
+                        meshPolygons.push_back(polygon);
+                    }
+                }
+                //break;
+            }
+
+
+        }
+
+        /** @brief Given an intial startEdge, do a BFS to identify the polygon
+        * that the edge is in. Also populates the halfEdgeQueue with the
+        */
+        bool BFSIdentifyPolygon(meshHalfEdgePtr startEdge
+            , std::shared_ptr<SplittablePolygon>& polygon, std::queue<meshHalfEdgePtr>& halfEdgeQueue
+            , const math::mat4& transform)
+        {
+            log::debug("->BFSIdentifyPolygon");
+            //polygonEdgeList : edges considered to be in the same polygon
+            std::vector<meshHalfEdgePtr> edgesInPolygon;
+
+            meshHalfEdgePtr nextEdge = nullptr;
+            meshHalfEdgePtr prevEdge = nullptr;
+
+            //startEdge may not form a triangle,we early out if this happens
+            if (!startEdge->AttemptGetTrianglesInEdges(nextEdge, prevEdge))
+            {
+                return false;
+            }
+
+
```

Comment:
nextEdge and prevEdge might have been changed by the AttemptGetTrianglesInEdges function